### PR TITLE
impl(generator): add rest stub factory

### PR DESCRIPTION
**Pull request by scotthart**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10296)

### CODE REVIEW

File: generator/internal/stub_factory_rest_generator.cc [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+       "google/cloud/options.h", "google/cloud/log.h"});
+  CcSystemIncludes({"memory"});
+
+  auto result = CcOpenNamespaces(NamespaceType::kInternal);
+  if (!result.ok()) return result;
+
+  // factory function implementation
+  CcPrint(R"""(
+std::shared_ptr<$stub_rest_class_name$>
+CreateDefault$stub_rest_class_name$(Options const& options) {
```
Comment (coryan) [inline generator/internal/stub_factory_rest_generator.cc:89]: The change suggested above would need to happen here, of course.

File: generator/internal/stub_factory_rest_generator.cc [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  Options auth_option;
+  if (!options.has<UnifiedCredentialsOption>()) {
+    auth_option.set<UnifiedCredentialsOption>(MakeGoogleDefaultCredentials());
+  })""");
+
+  CcPrint(R"""(
```
Comment (coryan) [inline generator/internal/stub_factory_rest_generator.cc:95]: I am not sure why you broke these into multiple `CcPrint()` calls?  Seems like a single raw string would do the trick?

File: google/cloud/common_options.h [hunk 3]
```diff
@@ -66,6 +66,7 @@ struct UserAgentProductsOption {
  *
  * - rpc
  * - rpc-streams
+ * - http Only applicable when using REST transport.
```
Comment (coryan) [inline google/cloud/common_options.h:69]: `http` already means something for `google/cloud/storage/`: it dumps every byte sent and received over the wire.  Why not just use `rpc` as before?

File: generator/integration_tests/tests/golden_kitchen_sink_rest_stub_factory_test.cc [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+namespace google {
+namespace cloud {
+namespace golden_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+using ::testing::HasSubstr;
+using ::testing::IsEmpty;
+
+TEST(GoldenKitchenSinkRestStubFactoryTest, DefaultStubWithoutLogging) {
```
Comment (coryan) [inline generator/integration_tests/tests/golden_kitchen_sink_rest_stub_factory_test.cc:32]: This (as you probably know) does not test that the metadata decorator is created.  Do you have plans to add that later (say in some kind of "test the full stack" test)?

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10296)* status: 0 of 15 files reviewed, 5 unresolved discussions (waiting on @coryan)

___
*[`generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub_factory.cc` line 38 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10296#-NHWdhqo8R6eRbauxwdW-r1-38:-NH_qhWIDOX5lc2-RMiv:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/ef44b3585443ac2b57320d166b5d529e81af574d/generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub_factory.cc#L38)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I think this would be simpler:

```
std::shared_ptr<GoldenKitchenSinkRestStub>
CreateDefaultGoldenKitchenSinkRestStub(Options const& opts) {
  Options options = opts;
    if (!options.has<UnifiedCredentialsOption>()) {
      options.set<UnifiedCredentialsOption>(MakeGoogleDefaultCredentials());
    }
    std::shared_ptr<GoldenKitchenSinkRestStub> stub =
      std::make_shared<DefaultGoldenKitchenSinkRestStub>(std::move(options));
...
```


</blockquote></details>

Done.
___
*[`generator/integration_tests/tests/golden_kitchen_sink_rest_stub_factory_test.cc` line 32 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10296#-NHWdhqo8R6eRbauxwd_-r1-32:-NH_qi1B8EwRyznVhSzO:b-bqqxc1) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/ef44b3585443ac2b57320d166b5d529e81af574d/generator/integration_tests/tests/golden_kitchen_sink_rest_stub_factory_test.cc#L32)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

This (as you probably know) does not test that the metadata decorator is created.  Do you have plans to add that later (say in some kind of "test the full stack" test)?
</blockquote></details>

Yes. I need a full stack to test a few things. The metadata decorator creation, honoring EndpointOption overrides, etc.
___
*[`generator/internal/stub_factory_rest_generator.cc` line 89 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10296#-NHWdhqo8R6eRbauxwdb-r1-89:-NH_r0u0EUQzoeFkIyQ3:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/ef44b3585443ac2b57320d166b5d529e81af574d/generator/internal/stub_factory_rest_generator.cc#L89)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

The change suggested above would need to happen here, of course.
</blockquote></details>

Done.
___
*[`generator/internal/stub_factory_rest_generator.cc` line 95 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10296#-NHWdhqo8R6eRbauxwdb-r1-95:-NH_r1m22Angc4KYmK_X:b-b8k4vw) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/ef44b3585443ac2b57320d166b5d529e81af574d/generator/internal/stub_factory_rest_generator.cc#L95)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I am not sure why you broke these into multiple `CcPrint()` calls?  Seems like a single raw string would do the trick?
</blockquote></details>

Combined.
___
*[`google/cloud/common_options.h` line 69 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10296#-NHWdhqo8R6eRbauxwdU-r1-69:-NH_r8NS9nKkF0xO2NjQ:b-t1l322) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/ef44b3585443ac2b57320d166b5d529e81af574d/google/cloud/common_options.h#L69)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

`http` already means something for `google/cloud/storage/`: it dumps every byte sent and received over the wire.  Why not just use `rpc` as before?
</blockquote></details>

Switched to rpc.

**Review by coryan — APPROVED** [review]
:lgtm:

> Reviewed 9 of 15 files at r1, 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10296#-:-NH_wfNNCjVFVBMBuK6l:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @scotthart)

**Pull request merged**